src/compiler/crystal/semantic/type_guess_visitor.cr

```diff
@@ -566,16 +566,31 @@ module Crystal
         return type if type.is_a?(PointerInstanceType)
       end
 
-      if type = guess_type_call_pointer_malloc_two_args(node)
-        return type
-      end
+      type = guess_type_call_pointer_malloc_two_args(node)
+      return type if type
 
       type = guess_type_call_lib_fun(node)
       return type if type
 
+      # If it's `obj.tap { ... }`, guess from `obj`
+      obj = is_tap_call?(node)
+      if obj
+        return guess_type(obj)
+      end
+
       nil
```

Comment:
Actually, you can. You can't if you declare the variable in the `if`, for example this won't work:

```crystal
foo(bar) if bar = some_expression
```

But I don't always use a suffix `if` whenever I can, in some cases I find it more clear not to use it.